File: tools/probes/s75-stacked-deletions.cs
#:project ../../src/FuzzyRegex/FuzzyRegex.csproj
// S75 (2026-09-20): the port's half of the owner's `(foobar){e}` case. The upstream half, and what
// the case is about, is in tools/probes/s75-stacked-deletions.py.
//
//     pattern  (foobar){e}
//     subject  xirefoabralfobarxie
//
// Two columns, because the demo does not draw the library's own deletion positions. A deletion is
// reported where the missing character would sit in a string that had every deletion put back, so
// the i-th deletion of a match is i past the place in the subject it happened at; DemoEngine.Edits
// subtracts that i so the page has a position in the string it slices (DemoEngine.cs, the
// `changes.Deletions[i] - i`). "library" below is what upstream is compared against, "demo" is what
// reaches the highlight, and the stacking the slice fixes is only visible in the second.
//
//     dotnet run tools/probes/s75-stacked-deletions.cs

using System.Globalization;
using Fuzzy.Text.RegularExpressions;

const string pattern = "(foobar){e}";
const string subject = "xirefoabralfobarxie";

Console.WriteLine($"pattern {pattern}  subject {subject}");
Console.WriteLine();

// PadRight/PadLeft rather than interpolation alignment specifiers: CSharpier writes those with a
// space after the comma and IDE0055 rejects the space, so a line using one satisfies neither of
// this repo's formatting gates (the same note is on tools/probes/aot-smoke-slow-patterns.cs).
Console.WriteLine("  #       span  text         s  i  d  library deletions -> demo deletions");

int number = 0;
foreach (Match match in FuzzyRegex.EnumerateMatches(subject, pattern))
{
    number++;
    FuzzyCounts counts = match.FuzzyCounts;
    FuzzyChanges changes = match.FuzzyChanges;

    int[] library = [.. changes.Deletions];
    int[] demo = [.. library.Select(static (at, i) => at - i)];

    string span = $"({match.Index},{match.Index + match.Length})";
    string counted =
        counts.Substitutions.ToString(CultureInfo.InvariantCulture).PadLeft(2)
        + " "
        + counts.Insertions.ToString(CultureInfo.InvariantCulture).PadLeft(2)
        + " "
        + counts.Deletions.ToString(CultureInfo.InvariantCulture).PadLeft(2);
    Console.WriteLine(
        number.ToString(CultureInfo.InvariantCulture).PadLeft(3)
            + "  "
            + span.PadLeft(9)
            + "  "
            + $"'{match.Value}'".PadRight(10)
            + "  "
            + counted
            + $"  [{string.Join(", ", library)}] -> [{string.Join(", ", demo)}]"
    );
}

File: tools/probes/s75-stacked-deletions.py
# S75 (2026-09-20): the owner's `(foobar){e}` case, which the demo drew as eleven `d` markers on
# top of each other.
#
#     pattern  (foobar){e}
#     subject  xirefoabralfobarxie
#
# The drawing was wrong and the engine was right, so this probe is the half that proves the second
# half: what upstream answers, span by span, with the errors each match spent. Run it before
# changing how the markers are drawn, because a marker row built against a wrong span count would
# look correct and be measuring the demo's own bug.
#
# `{e}` carries NO bound, so every position matches and the answer is long. The two rows the slice
# is about are the last two: a match of a single character that spent one substitution and five
# deletions, and an EMPTY match at the end of the subject that spent six. Six deletions in one
# place is six marks stacked at one x-position under the old drawing, and an empty match has no
# character to hang a highlight on at all.
#
# Pair: tools/probes/s75-stacked-deletions.cs runs the same case through this port.
#
#     python tools/probes/s75-stacked-deletions.py

import regex

PATTERN = '(foobar){e}'
SUBJECT = 'xirefoabralfobarxie'

print(f'regex {regex.__version__}')
print(f'pattern {PATTERN!r}  subject {SUBJECT!r}')
print()
print(f'{"#":>3}  {"span":>9}  {"text":<10}  {"s":>2} {"i":>2} {"d":>2}  fuzzy_changes')

compiled = regex.compile(PATTERN, regex.VERSION1)
for number, match in enumerate(compiled.finditer(SUBJECT), start=1):
    subs, inss, dels = match.fuzzy_changes
    substitutions, insertions, deletions = match.fuzzy_counts
    span = f'({match.start()},{match.end()})'
    print(
        f'{number:>3}  {span:>9}  {match.group()!r:<10}  '
        f'{substitutions:>2} {insertions:>2} {deletions:>2}  '
        f'subs={subs} ins={inss} dels={dels}'
    )
